### PR TITLE
Workaround su

### PR DESCRIPTION
**Pull request by faho**
## Description

As we've seen, some `su`s keep $USER and $HOME if the target user is root (I'm assuming that means uid == 0, so `su toor` would also be affected).

Since this seems to be the only case where $USER points to an existing but wrong user, we work around this explicit case by force-initializing $USER and $HOME.

Fixes issue #3916.

## TODOs:

- [N/A] Changes to fish usage are reflected in user documentation/manpages.
- [N/A] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.md

I'd like to see this tested on a system with multiple uid 0 users (e.g. the above-mentioned "toor").

### CODE REVIEW

File: src/env.cpp [hunk 1]
```diff
@@ -803,7 +803,14 @@ void env_init(const struct config_paths_t *paths /* or NULL */) {
 
     // Set up the USER and PATH variables
     setup_path();
-    setup_user(false);
+
+    // Some `su`s keep $USER (and $HOME)
```
Comment (krader1961) [inline src/env.cpp]: Nit: This comment can be reformatted to fit on fewer lines.

**Comment by krader1961** [comment]
LGTM modulo the minor nit about the comment split over more lines than necessary.

**Comment by zanchey** [comment]
If this is in place, can I launch fish with a different home directory as root?

**Comment by krader1961** [comment]
> If this is in place, can I launch fish with a different home directory as root?

No and I don't see any reasonable way to support that. Have you actually needed that capability? It would be nice to support it but not at the expense of continuing to have people be confused when they do `su root`.

**Comment by faho** [comment]
>Have you actually needed that capability?

Yes, we have, and I forgot about that (sorry @zanchey!).

One instruction in our issue template is

    sh -c 'env HOME=$(mktemp -d) fish'

This is to get a pristine fish without configuration, to help debugging. The way to support that again would be to add an explicit "--norc"-like option.

Without that, this should not be released, since it'd make some bug reports a major PITA.

I have a preliminary implementation (that just ignores $HOME), but I think we have an issue that might provide some insight into how exactly it should work. I'll try to find it.

**Comment by zanchey** [comment]
The fish source uses it in `tests/expansion.in`, for example.

I think we need to be careful here about solving a corner case and messing up the general case in a way that can't be overridden.

**Comment by krader1961** [comment]
Right, but if I'm reading the code correctly this change only makes that impossible for the root account. It should work fine for non-root users. So my question was have you needed that capability while logged in as root?

**Comment by zanchey** [comment]
Personally, not by the look of my history in interactive use. If we are using it in our scripts and recommendations why should others not?

**Comment by faho** [comment]
Wait a minute, this doesn't work in the first place, because `su` sets the $HOME!

```fish
$ env HOME=(mktemp -d) su
# echo $USER $HOME
alfa /root
```

From the man page for my particular implementation (the one from util-linux):

>For  backward  compatibility,  su  defaults to not change the current directory and to only set the environment variables HOME and SHELL (plus USER and LOGNAME if the target user is not root).

A quick google for `man su` finds a couple of different man pages that either also talk of setting $HOME or completely lack any info on what variables they set.

So I was actually mistaken about this applying to $HOME.

Now, there's the "-m" or "-p" or "--preserve-environment" option, which keeps $HOME, so we might want to just use $HOME. I make no guarantees that that does anything sensible without setting home to a temporary directory or something similar, though. But in that case the user made an explicit request that we should honor. If it breaks anything, they get to keep the pieces.

**Comment by krader1961** [comment]
Whether or not `$HOME` is set by `su` (when invoked without `-l`) is another implementation dependent inconsistency. GNU Linux has the `--preserve-environment` flag but BSD does not. And it's frankly not obvious why `$HOME` should ever be set by `su` to anything other than the home directory for the target account. But then the same could be said about the `$USER` env var. No doubt there is some scenario in which not having those two vars reflect reality after `su root` makes sense. But it does seem to cause more problems than benefit.

As for @zanchey's statement that "If we are using it in our scripts..." (referring to the env trick for debugging) I can find nowhere that we do this other than the *tests/expansion.in* unit test. And the fact that test would be broken by this change if run as root is not something that bothers me since you should never run our unit tests as root. Also, I can't find this trick in our documentation outside of the github issue template. Nonetheless, we probably should allow people to use the `env HOME=/someplace fish` trick as the root user. In which case @faho's latest addendum to this PR is what we want.

I vote we merge this PR in its present form. That is, munging `$USER` but not `$HOME`. That should "fix" most of the problem of doing `su root` rather than `su -l root`. It is possible it will cause problems for someone expecting `$USER` to not change but I can't muster much sympathy for them.

**Comment by faho** [comment]
>And it's frankly not obvious why $HOME should ever be set by su to anything other than the home directory for the target account. But then the same could be said about the $USER env var. No doubt there is some scenario in which not having those two vars reflect reality after su root makes sense.

I believe the idea was that $USER would reflect which user used `su` (in a certain sense which user "you really are"). Since the uid in this case is by definition 0, you could just check e.g. sh's $UID variable to figure out that you are root. $HOME would not help there, and since it's used to find config files that would cause even more issues.

I actually can't think of a case where that information is actually useful, though. Maybe for logging? E.g. "who made this change as the root user"? Which would also explain the $LOGNAME variable, which is what I'd recommend if you actually need this.

>But it does seem to cause more problems than benefit.

I certainly agree. Which is also why I'm not going the route of adding a $UID variable instead.

>I can find nowhere that we do this other than the tests/expansion.in unit test.

To test tilde-expansion. See https://github.com/fish-shell/fish-shell/blob/8b67a1b26f2525f7de1e2a64dee6ab48b72b0e33/tests/expansion.in#L97:

```fish
set expandedtilde (env HOME=$tmpdir/linkhome ../test/root/bin/fish -c 'echo ~')
if test $expandedtilde != $tmpdir/realhome
	echo '~ expands to' $expandedtilde ' - expected ' $tmpdir/realhome
end
```

>I vote we merge this PR in its present form. That is, munging $USER but not $HOME.

Yes, though I'll remove the second commit ("Get $HOME via uid"), since I'm not sure how that interacts with the root/toor thing, and I've not found which passwd entry `getpwuid` returns in that case (though I'd assume it's the first). It's also not necessary for this change - it was just a speculative optimization.

**Comment by faho** [comment]
>To test tilde-expansion.

Note: This could be replaced with `(set -lx HOME $tmpdir/linkhome; echo ~)`, unless of course we also want to test the `env HOME=` idiom.

**Comment by faho** [comment]
I've now merged this without the $HOME changes as 3fa5d6c79405c2f5b1299af5d487a7bc95b3a10e.

That won't impact the `env HOME` trick as it never touches $HOME, but it will fix `su -c fish`.
